SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/RPCService.cpp

```diff
@@ -0,0 +1,312 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/RPCs/RPCService.h"
+
+namespace SpatialGDK
+{
+RPCService::RPCService(const FSubView& InRemoteSubView, const FSubView& InLocalAuthSubView)
+	: RemoteSubView(&InRemoteSubView)
+	, LocalAuthSubView(&InLocalAuthSubView)
+{
+}
+
+void RPCService::AddRPCQueue(FName QueueName, RPCQueueDescription&& Desc)
+{
+	Queues.Add(QueueName, MoveTemp(Desc));
+}
+
+void RPCService::AddRPCReceiver(FName ReceiverName, RPCReceiverDescription&& Desc)
+{
+	Receivers.Add(ReceiverName, MoveTemp(Desc));
+}
+
+bool RPCService::HasReceiverAuthority(const RPCReceiverDescription& Desc, const EntityViewElement& ViewElement)
+{
+	return Desc.Authority == NoAuthorityNeeded || ViewElement.Authority.Contains(Desc.Authority);
+}
+
+bool RPCService::IsReceiverAuthoritySet(const RPCReceiverDescription& Desc, Worker_ComponentSetId ComponentSet)
+{
+	return Desc.Authority != NoAuthorityNeeded && ComponentSet == Desc.Authority;
+}
+
+void RPCService::ProcessUpdatesToSender(Worker_EntityId EntityId, ComponentSpan<ComponentChange> Updates)
+{
+	for (const ComponentChange& Change : Updates)
+	{
+		RPCReadingContext Ctx;
+		Ctx.EntityId = EntityId;
+		Ctx.ComponentId = Change.ComponentId;
+		if (Change.Type == ComponentChange::COMPLETE_UPDATE)
+		{
+			Ctx.Fields = Schema_GetComponentDataFields(Change.CompleteUpdate.Data);
+		}
+		else
+		{
+			Ctx.Update = Change.Update;
+			Ctx.Fields = Schema_GetComponentUpdateFields(Change.Update);
+		}
+
+		for (auto& QueueEntry : Queues)
+		{
+			if (QueueEntry.Value.Sender->GetComponentsToReadOnUpdate().Contains(Ctx.ComponentId))
+			{
+				QueueEntry.Value.Sender->OnUpdate(Ctx);
+			}
+		}
+	}
+}
+
+void RPCService::AdvanceSenderQueues()
+{
+	for (const EntityDelta& Delta : LocalAuthSubView->GetViewDelta().EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			ProcessUpdatesToSender(Delta.EntityId, Delta.ComponentUpdates);
+			ProcessUpdatesToSender(Delta.EntityId, Delta.ComponentsRefreshed);
+			break;
+		}
+		case EntityDelta::ADD:
+		{
+			const EntityViewElement& ViewElement = LocalAuthSubView->GetView().FindChecked(Delta.EntityId);
+			for (auto& QueueEntry : Queues)
+			{
+				if (ViewElement.Authority.Contains(QueueEntry.Value.Authority))
+				{
+					QueueEntry.Value.Queue->OnAuthGained(Delta.EntityId, ViewElement);
+					QueueEntry.Value.Sender->OnAuthGained(Delta.EntityId, ViewElement);
+				}
+			}
+		}
+		break;
+		case EntityDelta::REMOVE:
+			for (auto& QueueEntry : Queues)
```

Comment:
very minor: indentation weirdness

Reply:
What is weird ? Looks ok on my side.

Reply:
I guess I was expecting either braces or the for loop aligned with the lhs of case

Reply:
Yeah indentation rules for switch/case are non-intuitive.